### ISSUE TITLE
Test fails to use new contract addresses if evm_revert isn't supported.

### ISSUE THREAD

**Issue by no2chem** [trufflesuite/truffle#3023]
## Issue

When running `truffle test` against a client that doesn't support `evm_revert`, Truffle will try to reset by redeploying the contracts between test suites.

https://github.com/trufflesuite/truffle/blob/0effb767df26dca7aab0b1247de30fdf5b127be9/packages/core/lib/testing/testrunner.js#L81 

However, contracts referred to by the new test suite (via `artifact.require`) point to the previously deployed contracts instead of the new ones. This causes issues for tests which expect the state of the contracts to be reset in the new test suite.

As far as I can tell by debugging through this, even though the contracts are redeployed in testrunner::deploy:

https://github.com/trufflesuite/truffle/blob/0effb767df26dca7aab0b1247de30fdf5b127be9/packages/core/lib/testing/testrunner.js#L72

The artifacts are loaded when mocha *first* loads the test files, right after the inital deploy. testrunner::deploy doesn't reload the artifacts, leading to the problem. The new redeployed contract addresses need to be reinjected to the already loaded contracts somehow.

## Steps to Reproduce

On a client the doesn't support `evm_revert` run a minimal test suite:

contracts/Increment.sol
```sol
pragma solidity >=0.5.0;

contract Increment {
    uint256 count;
    
    function increment() public {
        count++;
    }
    
    function getCount() view public returns (uint256) {
        return count;
    }
```

tests/test1.js
```js
const Increment = artifacts.require("Increment");

contract("First test", async accounts => {
     it("should increment to one", async() => {
            let deployed = await Increment.deployed();
            await deployed.increment.sendTransaction();
            console.log(`Address: ${deployed.address}`);
            expect(await deployed.getCount()).to.equal(1);
     });
});
```
tests/test2.js
```js
const Increment = artifacts.require("Increment");

contract("Second test", async accounts => {
     it("should increment to one", async() => {
            let deployed = await Increment.deployed();
            await deployed.increment.sendTransaction();
            // ISSUE 1: Address is the same as previous suite!
            console.log(`Address: ${deployed.address}`);
            // ISSUE 2: TEST FAILURE - returns 2 instead of 1 due to previous state
            expect(await deployed.getCount()).to.equal(1);
     });
});
```

## Expected Behavior

The second test suite results in the issues described above.

## Actual Results

The second test suite runs with the previously deployed contracts as the first test suite, causing the issues described above.

## Environment

* Operating System: macOS catalina
* Ethereum client: test client which doesn't support eth_revert ethrpc call.
* Truffle version (`truffle version`): 5.1.24
* node version (`node --version`): 13.3.0
* npm version (`npm --version`): 6.13.4

**Comment by haltman-at** [comment]
This is a duplicate of #2735, but maybe we should leave it up for the extra detail. :)

**Comment by no2chem** [comment]
Thanks for pointing to the duplicate issue, @haltman-at !

I'm happy to take a stab at this. It looks like the easiest way to resolve this is just to create a new Mocha instance for each test file. This might slightly affect test behavior though...

**Comment by haltman-at** [comment]
Hm, I would hope there would be a better way than that.  Note that things are supposed to reset between each `contract()`, rather than between each file.  But, yeah... any solution would be nice.  If you can come up with something maybe we can figure out a way to do it better!  None of us have yet had time to take a serious stab at this...

The most relevant files here, by the way, are [these](https://github.com/trufflesuite/truffle/blob/develop/packages/core/lib/test.js) [two](https://github.com/trufflesuite/truffle/blob/develop/packages/core/lib/testing/testrunner.js).

I wonder if the real problem here is that `global.artifacts` isn't getting reset when we do a reset?  But the `testResolver` has its cache turned off, so that shouldn't matter, seems to me?  Yeah, I remain pretty confused as to why this is happening.  But I'm not too familiar with these systems, unfortunately.

**Comment by no2chem** [comment]
I think the problem is that each .js file is loaded only once, when mocha is initialized. So the top-level `const Contract = artifacts.require('contract')` is called only right after the first deployment.

Resetting `global.artifacts` doesn't help, as the top level contract requires in each suite aren't called again.

**Comment by haltman-at** [comment]
Oh, duh, good point.  Yeah, the fact that such things typically go at top-level makes things trickier.  I guess I don't know what to do about that offhand.  Thanks for looking at this!

**Comment by haltman-at** [comment]
OK, I think @gnidan and I have come up with a way to handle this other than resetting Mocha.  It's fairly hacky, but it should work.  I'm going to give it a try and hopefully I should have a PR up in a bit.
